README.md
# imageZIP

This is a small python package to archive (encrypt) files and directories into a single image file. 

This project is made just for fun, it has no practical use, output "ZIP" image will always have the same or greater file size than it's content. Also, this is not easy-to-scan like QR codes.

Algorithm is pretty easy. If folder to zip is chosen, it will iterate through all folders and files inside it. If folder is found, it will write relative path of folder in bytes and add special folder delimiter at the end. If file is found, it will write it's relative path and name in bytes, add special file delimiter and then write all it's content in bytes with additional second file delimiter. When bytes array is ready to go, it will build an image, where individual pixels representing particular byte number.

## Example

![demo image](.github/demo_zip.png)

This is an "imageZIP archive", containing this whole "imageZIP" repository.
 
It was created with ```imageZIP.zip("some path/imageZIP/", scale=4)```.

You can actually download this image and unzip it with ```imageZIP.unzip("some path/demo_zip.png", scale=4)```, it will create "imageZIP" folder with all repository files.

## How to use

To download this package, you can use ```pip install imageZIP```.

Then go to python terminal and write ```import imageZIP```.

Then you can use 2 available functions:

- ```imageZIP.zip(path: str, scale: int = 1, color_mode: str = 'heat')```;
- ```imageZIP.unzip(path: str, scale: int = 1, color_mode: str = 'heat')```.

Where:
- ```path``` - path (```str```) to a single file or whole directory to "zip", or image file to "unzip";
- ```scale```* - scale (```int```) of input or output image file. Default: ```1```;
- ```color_mode```* - color mode (```str```) of input or output image file. Default: ```"heat"```.

> *not necessary attributes, which can be omitted.

#### Notes:

- For ```zip``` and ```unzip``` you have to provide same ```scale``` and ```color_mode``` settings;
- Available ```color_mode```s are: ```bw```, ```heat```, ```heat_toxic```, ```rainbow```, ```red```, ```blue```;
- ```scale``` is have to be ```int >= 1```;
- ```path``` can be absolute or relative;
- If a good rectangle image can't be created, output image will be a square with some empty pixels at the end.


CHANGELOG.md
Change Log
=========================

3.1 (02.12.2020)
-------------------------
- Fixed "OSError" while copy-pasting path from windows explorer;
- Removed unused "file_chosen" attribute.

3 (27.11.2020)
-------------------------
- Linux path fix for an output image file;
- New color modes: "rainbow", "red", "blue";
- "heat_map" and "heat_map_toxic" color modes is now can be used with more shorter "heat" and "heat_toxic" names;
- Docstrings.
- Other minor fixes.

2.1 (25.11.2020)
-------------------------
- Small linux fix.

2 (25.11.2020)
-------------------------
- Fixed working on Linux.

1 (10.11.2020)
-------------------------
#### Release.

tests.py
from imageZIP import imageZIP
import unittest
import os
import shutil


class TestBehavior(unittest.TestCase):
    def test_file(self):
        original_file_name = 'tests/test file ☆.txt'
        zip_path = 'tests/test file ☆.txt_zip.png'
        with open(original_file_name, 'r', encoding='UTF-8') as file:
            original_file = file.read()

        imageZIP.zip(original_file_name)
        self.assertTrue(os.path.exists(zip_path))
        self.assertFalse(os.path.exists('test file ☆.txt_zip.png'))

        os.remove(original_file_name)

        imageZIP.unzip(zip_path)
        with open(original_file_name, 'r', encoding='UTF-8') as file:
            unzipped_file = file.read()
        self.assertTrue(os.path.exists(original_file_name))
        self.assertEqual(unzipped_file, original_file)

        os.remove(zip_path)

    def test_blank_file(self):
        original_file_name = 'tests/test folder/0'
        zip_path = 'tests/test folder/0_zip.png'
        with open(original_file_name, 'r', encoding='UTF-8') as file:
            original_file = file.read()

        imageZIP.zip(original_file_name)
        self.assertTrue(os.path.exists(zip_path))
        self.assertFalse(os.path.exists('0_zip.png'))

        os.remove(original_file_name)

        imageZIP.unzip(zip_path)
        with open(original_file_name, 'r', encoding='UTF-8') as file:
            unzipped_file = file.read()
        self.assertTrue(os.path.exists(original_file_name))
        self.assertEqual(unzipped_file, original_file)

        os.remove(zip_path)

    def test_binary_file(self):
        original_file_name = 'tests/test folder/inner folder/image.jpg'
        zip_path = 'tests/test folder/inner folder/image.jpg_zip.png'
        with open(original_file_name, 'rb') as file:
            original_file = file.read()

        imageZIP.zip(original_file_name)
        self.assertTrue(os.path.exists(zip_path))
        self.assertFalse(os.path.exists('image.jpg_zip.png'))

        os.remove(original_file_name)

        imageZIP.unzip(zip_path)
        with open(original_file_name, 'rb') as file:
            unzipped_file = file.read()
        self.assertTrue(os.path.exists(original_file_name))
        self.assertEqual(unzipped_file, original_file)

        os.remove(zip_path)

    def test_folder(self):
        original_folder_name = 'tests/test folder/'
        zip_path = 'tests/test folder_zip.png'
        path_empty_file = os.path.join(original_folder_name, '0')
        with open(path_empty_file, 'rb') as file:
            empty_file = file.read()
        path_image_file = os.path.join(original_folder_name, 'inner folder/image.jpg')
        with open(path_image_file, 'rb') as file:
            image_file = file.read()

        imageZIP.zip(original_folder_name)
        self.assertTrue(os.path.exists(zip_path))
        self.assertFalse(os.path.exists('test folder_zip.png'))

        shutil.rmtree(original_folder_name, ignore_errors=True)

        imageZIP.unzip(zip_path)
        with open(path_empty_file, 'rb') as file:
            empty_file_new = file.read()
        self.assertEqual(empty_file_new, empty_file)
        with open(path_image_file, 'rb') as file:
            image_file_new = file.read()
        self.assertEqual(image_file_new, image_file)
        self.assertFalse(os.path.exists('test folder\\0') and os.path.isdir('test folder\\0'))

        os.remove(zip_path)

    def DISABLED_test_abs_path_folder(self):
        if os.name == 'nt':
            os.mkdir('C:/test_imageZIP')
            os.mkdir('C:/test_imageZIP/test folder/')
            original_folder_name = 'C:/test_imageZIP/test folder/'
            zip_path = 'C:/test_imageZIP/test folder_zip.png'
        else:
            os.mkdir('/test_imageZIP/')
            os.mkdir('/test_imageZIP/test folder/')
            original_folder_name = '/test_imageZIP/test folder/'
            zip_path = '/test_imageZIP/test folder_zip.png'
        path_file = os.path.join(original_folder_name, '0')
        with open(path_file, 'w') as file:
            file.write('testing 123')
        with open(path_file, 'rb') as file:
            text_file = file.read()

        imageZIP.zip(original_folder_name)
        self.assertTrue(os.path.exists(zip_path))
        self.assertFalse(os.path.exists('test folder_zip.png'))

        if os.name == 'nt':
            shutil.rmtree(original_folder_name, ignore_errors=True)
        else:
            shutil.rmtree(original_folder_name, ignore_errors=True)

        imageZIP.unzip(zip_path)
        with open(path_file, 'rb') as file:
            text_file_new = file.read()
        self.assertEqual(text_file_new, text_file)

        if os.name == 'nt':
            shutil.rmtree('C:/test_imageZIP', ignore_errors=True)
        else:
            shutil.rmtree('/test_imageZIP/', ignore_errors=True)

    def test_unicode_LEFT_TO_RIGHT_EMBEDDING(self):
        path = '‪C/Users/An/Un/Existing/Route123123/'

        with self.assertRaises(FileNotFoundError):
            imageZIP.zip(path)


class TestSettings(unittest.TestCase):
    def _test_with_settings(self, *args, **kwargs):
        original_file_name = 'tests/test file ☆.txt'
        zip_path = 'tests/test file ☆.txt_zip.png'
        with open(original_file_name, 'r', encoding='UTF-8') as file:
            original_file = file.read()

        imageZIP.zip(original_file_name, *args, **kwargs)
        self.assertTrue(os.path.exists(zip_path))
        self.assertFalse(os.path.exists('test file ☆.txt_zip.png'))

        os.remove(original_file_name)

        imageZIP.unzip(zip_path, *args, **kwargs)
        with open(original_file_name, 'r', encoding='UTF-8') as file:
            unzipped_file = file.read()
        self.assertTrue(os.path.exists(original_file_name))
        self.assertEqual(unzipped_file, original_file)

        os.remove(zip_path)

    def test_scale(self):
        with self.assertRaises(SystemError):
            self._test_with_settings(scale=0)
        with self.assertRaises(TypeError):
            self._test_with_settings(scale=1.2)
        self._test_with_settings(scale=4)

    def test_colors(self):
        with self.assertRaises(AttributeError):
            self._test_with_settings(color_mode='random123123')
        self._test_with_settings(color_mode='bw')
        self._test_with_settings(color_mode='rainbow')
        self._test_with_settings(color_mode='RaInBow')  # Test char cases
        self._test_with_settings(color_mode='red')
        self._test_with_settings(color_mode='blue')
        self._test_with_settings(color_mode='heat')
        self._test_with_settings(color_mode='heat_toxic')
        self._test_with_settings(color_mode='heat_map')
        self._test_with_settings(color_mode='heat_map_toxic')


if __name__ == '__main__':
    unittest.main()


imageZIP/imageCrypt.py
from PIL import Image
import io as io
import os
import math


def _calculate_color(num: int, color_mode: str):
    if color_mode == 'bw':
        return num
    elif color_mode == 'heat_map':
        # original num is always be 0 >= num <= 255 (it's actually 256 different bytes)
        # by multiplying num, we are getting wider range of colors
        num *= 4
        # all pallets multiplied and divided by num should be equal 256 to store right amount of data
        # palettes is a ranges of colors to store data in
        palette1 = 256
        # color_min and color_max is a color range of palette1
        # Don't forget that 0 counts, so full range from color_min to color_max is 256, not 255
        color_min = 0

    elif color_mode == 'heat_map_better':
        num *= int(3.5)
        palette1 = 224
        color_min = 0

    else:
        raise AttributeError(f'"{color_mode}" color_mode does not exist.')

    color_max = color_min + palette1 - 1
    palette2 = palette1 * 2
    palette3 = palette2 + palette1
    palette4 = palette3 + palette1

    if num <= palette1:
        return color_min, color_min + num, color_max
    elif num <= palette2:
        num -= palette1
        return color_min, color_max, color_max - num
    elif num <= palette3:
        num -= palette2
        return color_min + num, color_max, color_min
    elif num <= palette4:
        num -= palette3
        return color_max, color_max - num, color_min


def _calculate_sizes(multiplier, perimeter):
    while multiplier > 0:
        a = perimeter / multiplier
        if a % 1 == 0:
            b = perimeter / a
            return int(min([a, b])), int(max([a, b]))
        multiplier -= 1


def draw_bytes_on_image(bts, color_mode, scale):
    bts_len = len(bts)
    bt_index = 0

    size_original = math.ceil(math.sqrt(bts_len))
    size_x, size_y = _calculate_sizes(size_original, bts_len)
    if color_mode == 'bw':
        color_PIL = 'L'
    else:
        color_PIL = 'RGB'
    img = Image.new(color_PIL, (size_x * scale, size_y * scale), color='white')

    for row in range(size_y):
        for col in range(size_x):
            try:
                color = _calculate_color(num=bts[bt_index], color_mode=color_mode)
            except IndexError:
                break
            x, y = col * scale, row * scale
            for i in range(scale):
                for j in range(scale):
                    img.putpixel((x + j, y + i), color)
            bt_index += 1
    return img


def _encrypt_save(overwrite, path, img):
    if overwrite:
        os.remove(path)
    img.save(path + '.png')


def encrypt_file(**kwargs):
    path = kwargs['path']
    scale = kwargs['scale']
    mode = kwargs['crypt_mode']
    overwrite = kwargs['overwrite']
    color_mode = kwargs['color_mode']

    with open(path, 'rb') as file:
        bts = file.read()

    if mode == 'normal':
        img = draw_bytes_on_image(bts, color_mode, scale)
        _encrypt_save(overwrite, path, img)
    elif mode == 'zip':
        print(path)
    else:
        raise AttributeError('Mode is not defined.')


def decrypt_file(**kwargs):
    path = kwargs['path']
    scale = kwargs['scale']
    overwrite = kwargs['overwrite']
    color_mode = kwargs['color_mode']

    with open(path, 'rb') as file:
        image = file.read()
    image = Image.open(io.BytesIO(image))

    bytesDict = {}
    for i in range(256):
        bytesDict[str(_calculate_color(i, color_mode=color_mode))] = bytes([i])

    bts = []

    for row in range(int(image.height / scale)):
        for col in range(int(image.width / scale)):
            try:
                bts.append(bytesDict[str(image.getpixel((col * scale, row * scale)))])
            except KeyError:
                raise RuntimeError("Can't decode an image, settings can be incorrect or image is distorted.")

    original_path, _ = os.path.splitext(path)
    if overwrite:
        file = open(original_path, 'wb')
    else:
        without_extension, extension = os.path.splitext(original_path)
        file = open(without_extension + '_decrypted' + extension, 'wb')
    file.writelines(bts)
    file.close()


def _crypt_with_mode(mode, **kwargs):
    if mode == 'encrypt':
        return encrypt_file(**kwargs)
    elif mode == 'decrypt':
        decrypt_file(**kwargs)
    else:
        raise AttributeError('imageCrypt mode is not defined.')


def crypt(path, bts_zip, crypt_mode, **kwargs):
    if os.path.isdir(path):
        for name in os.listdir(path):
            path_to_name = os.path.join(path, name)
            if os.path.isfile(path_to_name) or crypt_mode == 'zip':
                res = _crypt_with_mode(path=path_to_name, crypt_mode=crypt_mode, **kwargs)
                if crypt_mode == 'zip':
                    bts_zip.append(res)
            else:
                crypt(path_to_name, **kwargs)  # Recursion to inner folders
    else:
        _crypt_with_mode(path=path, crypt_mode=crypt_mode, **kwargs)


def encrypt(path: str,
            scale: int = 1,
            mode: str = 'zip',
            overwrite: bool = False,
            color_mode: str = 'heat_map'):
    """
    Encrypt a file or whole directories into an image file.
    :param path: Path to a file or directory.
    :param scale: Scale of output image. Increase for more clear image. Default: 1.
    :param mode: Encryption mode. "zip" is for archiving files and whole directories to one image file.
    This image file will store all the folders, files data, their names and relative locations.
    "normal" mode is for encrypting only file's data, if directory is specified, it will encrypt all files inside it,
    including all the inner directories files, but folders itself will stay the same.
    :param overwrite: Overwrite original file with encrypted image file or create a new image file with the same folder
    as original. Default: False.
    :param color_mode: Color mode of output image. Default: "heat_map".
    :param enforce_compact: If image can't be built as square or a rectangle, but only single line,
    it will build as square with some empty bytes at the end as white color (except "bw" mode).
    You can enforce it to be compact in any case. Default: False.
    """
    bts_zip = []
    crypt(path=path,
          mode='encrypt',
          scale=scale,
          crypt_mode=mode,
          overwrite=overwrite,
          color_mode=color_mode,
          bts_zip=bts_zip)
    print(path)
    _encrypt_save(overwrite, path, draw_bytes_on_image(bts_zip, scale=scale, color_mode=color_mode))


def decrypt(path: str,
            scale: int = None,
            mode: str = 'zip',
            overwrite: bool = False,
            color_mode: str = None):
    """
    Decrypt an image file or whole directories image files into an original file.
    :param path: Path to a file or directory. If directory is specified, it will decrypt all files inside it,
    including all the inner directories files.
    :param scale: Scale of input image. It's important to specify original scale.
    If scale is not specified, decryptor will try to find out original scale on it's own. Default: None.
    :param mode: Decryption mode. "zip" is for archiving files and whole directories to one image file.
    This image file is storing all the folders, files data, their names and relative locations.
    "normal" mode is for decrypting only file's data, if directory is specified, it will decrypt all files inside it,
    including all the inner directories files, but folders itself will stay the same.
    :param overwrite: Overwrite encrypted image file with decrypted original file or create a new original file in
    the same folder. Default: False.
    :param color_mode: Color mode of input image. It is also important to specify original color_mode.
    If color_mode is not specified, decryptor will try to find out original color_mode on it's own. Default: None.
    """
    crypt(path=path,
          mode='decrypt',
          scale=scale,
          crypt_mode=mode,
          overwrite=overwrite,
          color_mode=color_mode,
          bts_zip=[])


imageZIP/imageZIP.py
from PIL import Image
import io
import os
import math
from sys import platform

Image.MAX_IMAGE_PIXELS = None
ON_WINDOWS = platform == 'win32'


def _calculate_color(num, color_mode: str):
    """
    Calculate color for input byte, represented as 0<=int<=255 (or special float code)
    """

    color_max = 255

    if type(num) == float:
        if num == 0.:
            return 255, 250, 255
        elif num == 1.:
            return 5, 0, 5

    if color_mode in ['heat_toxic', 'heat_map_toxic']:
        # original num is always be 0 <= num <= 255 (it's actually 256 different bytes)
        # by multiplying num, we are getting wider range of colors
        num *= 4
        # all pallets multiplied and divided by num should be equal 256 to store right amount of data
        # palettes is a ranges of colors to store data in
        palette1 = 256
        # color_min and color_max is a color range of palette1
        # Don't forget that 0 counts, so full range from color_min to color_max is 256, not 255
        color_min = 0

    elif color_mode in ['heat', 'heat_map']:
        num *= 3
        palette1 = 192
        color_min = 20
    elif color_mode == 'bw':
        return num, num, num
    elif color_mode == 'rainbow':
        num *= 5
        palette1 = 183  # imperfect, actually 182.85
        color_min = 20
    elif color_mode == 'red':
        return color_max, color_max - num, color_max - num
    elif color_mode == 'blue':
        return color_max - num, color_max - num, color_max
    else:
        raise AttributeError(f'"{color_mode}" color_mode does not exist.')

    color_max = color_min + palette1 - 1
    palette2 = palette1 * 2
    palette3 = palette2 + palette1

    if color_mode == 'rainbow':
        palette4 = palette3 + palette1
        palette5 = palette4 + palette1
        palette6 = palette5 + palette1
        if num < palette1:
            return color_max, color_max - num, color_max - num
        if num < palette2:
            num -= palette1
            return color_max, color_min + num, color_min
        elif num < palette3:
            num -= palette2
            return color_max - num, color_max, color_min
        elif num < palette4:
            num -= palette3
            return color_min, color_max, color_min + num
        elif num < palette5:
            num -= palette4
            return color_min, color_max - num, color_max
        elif num < palette6:
            num -= palette5
            return color_min + num, color_min, color_max
        else:
            num -= palette6
            return color_max - num, color_min, color_max - num
    else:
        if num < palette1:
            return color_min, color_min + num, color_max
        elif num < palette2:
            num -= palette1
            return color_min, color_max, color_max - num
        elif num < palette3:
            num -= palette2
            return color_min + num, color_max, color_min
        else:
            num -= palette3
            return color_max, color_max - num, color_min


def _calculate_sizes(multiplier, perimeter):
    """
    Calculate sizes of final zip image
    """

    mult_orig = multiplier
    mult_stop = mult_orig / 2.5
    while multiplier > mult_stop:
        a = perimeter / multiplier
        if a % 1 == 0:
            b = perimeter / a
            return int(min([a, b])), int(max([a, b]))
        multiplier -= 1
    return mult_orig, mult_orig


def decrypt_image(path, color_mode, scale):
    """
    Unzips files and directories from an image file (path)
    """

    path_to_save = '/'.join(os.path.split(path)[:-1])

    if not ON_WINDOWS:
        path = path.replace(os.sep, '/').replace('\\', '/')

    with open(path, 'rb') as file:
        img = file.read()
    img = Image.open(io.BytesIO(img))

    bytesDict = {}
    for i in range(256):
        bytesDict[str(_calculate_color(i, color_mode=color_mode))] = bytes([i])
    bytesDict[str(_calculate_color(0., color_mode))] = 0  # Folder delimiter
    bytesDict[str(_calculate_color(1., color_mode))] = 1  # File delimiter
    bytesDict[str((230, 235, 230))] = 2  # End of file delimiter

    end_pixel = False
    interval = []
    decrypted_file_path = False

    # Not-so-good pile of code, redesign is needed
    for row in range(int(img.height / scale)):
        for col in range(int(img.width / scale)):
            try:
                bt = bytesDict[str(img.getpixel((col * scale, row * scale)))]
            except KeyError:
                raise RuntimeError("Can't unzip from an image, settings can be incorrect or image is distorted.")
            if type(bt) == int:
                if bt == 1:
                    if not decrypted_file_path:
                        decrypted_file_path = b''.join(interval).decode('utf-8')
                        interval = []
                        continue
                    else:
                        new_path = os.path.join(path_to_save, decrypted_file_path)

                        if not ON_WINDOWS:
                            new_path = new_path.replace(os.sep, '/').replace('\\', '/')

                        with open(new_path, 'wb') as file:
                            file.writelines(interval)

                        interval = []
                        decrypted_file_path = None
                elif bt == 0:
                    decrypted_folder_path = b''.join(interval).decode('utf-8')

                    path_joined = os.path.join(path_to_save, decrypted_folder_path)

                    if not ON_WINDOWS:
                        path_joined = path_joined.replace(os.sep, '/').replace('\\', '/')

                    if not os.path.exists(path_joined):
                        os.mkdir(path_joined)
                    interval = []
                elif bt == 2:
                    end_pixel = True
                    break
            else:
                interval.append(bt)
        if end_pixel:
            break


def draw_bytes_as_image(bts, color_mode, scale, **kwargs):
    """
    Build final zip image file from bytes
    """

    bts_len = len(bts)
    bt_index = 0

    size_original = math.ceil(math.sqrt(bts_len))
    size_x, size_y = _calculate_sizes(size_original, bts_len)
    color_PIL = 'RGB'
    img = Image.new(color_PIL, (size_x * scale, size_y * scale), color='#e6ebe6')

    for row in range(size_y):
        for col in range(size_x):
            try:
                color = _calculate_color(num=bts[bt_index], color_mode=color_mode)
            except IndexError:
                break
            x, y = col * scale, row * scale
            img_block = Image.new(color_PIL, (scale, scale), color=color)
            img.paste(img_block, (x, y))
            bt_index += 1
    return img


def img_save(img, path, **kwargs):
    """
    Save final image
    """

    path = os.path.normpath(path)

    if not ON_WINDOWS:
        path = path.replace(os.sep, '/').replace('\\', '/')

    img.save(path + '_zip.png')


def encrypt_obj(path_base, path_relative, **kwargs):
    """
    Encrypt file or directory
    """

    path = os.path.join(path_base, path_relative)

    bts = [b for b in bytes(''.join([char for char in path_relative]), encoding='UTF-8')]

    if os.path.isdir(path):
        bts.append(0.)

        return bts
    else:
        bts.append(1.)

        if not ON_WINDOWS:
            path = path.replace(os.sep, '/').replace('\\', '/')

        with open(path, 'rb') as file:
            file_bytes = file.read()
        [bts.append(b) for b in file_bytes]
        bts.append(1.)
        return bts


def encrypt(path_base, path_relative, bts_zip, **kwargs):
    """
    Recursive function, which is encrypting all the inner files and directories
    """
    path = os.path.join(path_base, path_relative)
    if os.path.isdir(path):
        bts_zip += encrypt_obj(path_base=path_base, path_relative=path_relative, **kwargs)
        for name in os.listdir(path):
            path_to_name = os.path.join(path_relative, name)
            encrypt(path_base=path_base, path_relative=path_to_name, bts_zip=bts_zip, **kwargs)  # Recursion
    else:
        bts_zip += encrypt_obj(path_base=path_base, path_relative=path_relative, **kwargs)


def encrypt_hub(path, **kwargs):
    """
    Whole encryption process for file or directory
    """
    bts_zip = []
    split = os.path.split(os.path.normpath(path))
    path_base = '/'.join(split[:-1])
    path_relative = split[-1]
    encrypt(path_base=path_base, path_relative=path_relative, bts_zip=bts_zip, **kwargs)
    img = draw_bytes_as_image(bts=bts_zip, **kwargs)
    img_save(img=img, **kwargs, path=path)


def validate_path(path):
    if path[0] == u'\u202a':
        return path[1:]
    return path


# EXPORTED FUNCTIONS

def zip(path: str,
        scale: int = 1,
        color_mode: str = 'heat'):
    """
    Encrypt (zip) chosen file or whole directory into a single image file.
    :param path: Path to a single file or whole directory to "zip". Can be absolute or relative.
    :param scale: Scale of output image file. Default: 1;
    :param color_mode: Color mode of output image file.
    Available color_mode's are: "bw", "heat", "heat_toxic", "rainbow", "red", "blue".
    Default: "heat".
    """

    color_mode = color_mode.lower()

    path = validate_path(path)

    encrypt_hub(path=path, scale=scale, color_mode=color_mode)


def unzip(path: str,
          scale: int = 1,
          color_mode: str = 'heat'):
    """
    Decrypt (unzip) files from an image file.
    :param path: Path to an image file to "unzip". Can be absolute or relative.
    :param scale: Scale of input image file. Default: 1;
    :param color_mode: Color mode of input image file.
    Available color_mode's are: "bw", "heat", "heat_toxic", "rainbow", "red", "blue".
    Default: "heat".
    """

    color_mode = color_mode.lower()

    path = validate_path(path)

    decrypt_image(path=path, scale=scale, color_mode=color_mode)
